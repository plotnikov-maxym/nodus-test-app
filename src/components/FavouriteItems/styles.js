export const styles = ({spacing: {unit}}) => ({
  root: {
    display: "flex",
    flexDirection: "column",
    alignItems: "flex-start",
  },
  flexRow: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
  },
  formControl: {
    margin: unit * 3,
  },
  group: {
    margin: `${unit}px 0`,
  },
  button: {
    margin: `${unit * 2}px 0 0`,
  },
  deleteButton: {
    marginLeft: unit * 2,
  },
});
