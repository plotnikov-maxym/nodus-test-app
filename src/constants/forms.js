export const ITEMS = [
  {
    value: 'arthroplasty',
    label: 'Total Hip Arthroplasty Revision',
  },
  {
    value: 'anterior',
    label: 'Anterior Lumbar Interbody Fusion Infection Rate',
  }];
