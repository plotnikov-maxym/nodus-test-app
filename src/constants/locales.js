export const locale = {
  WELCOME: "Welcome to Nodus research",
  SEARCH: "Search",
  SEARCH_IN_PUBMED: "Search in PubMed",
  TOO_SHORT_SEARCH_QUERY: "Please enter more than 3 characters for search.",
  ENTER_QUERY: "Please enter Your query to search.",
  NO_MATCHES: "No matches found.",
  FAVOURITE_ITEMS: "My favourite search items",
  DETAILS: "Details",
  ADD_TO_FAVOURITES: "Add to favourites",
  NO_FAVOURITES:
    "You don't have any favourites at the moment. You can add them at any time.",
};
