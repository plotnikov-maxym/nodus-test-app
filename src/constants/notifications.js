export const NOTIFICATION_TYPES = Object.freeze(["error", "info"]);
