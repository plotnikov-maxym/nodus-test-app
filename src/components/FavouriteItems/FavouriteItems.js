import React from "react";
import {object} from "prop-types";
import Typography from "@material-ui/core/Typography";
import Radio from "@material-ui/core/Radio";
import FormControl from "@material-ui/core/FormControl";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";
import DeleteIcon from "@material-ui/icons/Delete";
import {withStyles} from "@material-ui/core/styles";
import {locale} from "../../constants/locales";
import {styles} from "./styles";

class FavouriteItems extends React.Component {
  state = {
    value: "",
  };

  handleChange = event => {
    console.log(event.target.value);
    this.setState({value: event.target.value});
  };

  handleSubmit = event => {
    event.preventDefault();
    console.log(event);
  };

  deleteItem = item => {
    console.log(item);
  };

  render() {
    const {classes, heading, items} = this.props;

    return (
      <form
        className={classes.root}
        onSubmit={this.handleSubmit}
        noValidate
        autoComplete="off"
      >
        <div className={classes.formControl}>
          <Typography variant="h5" gutterBottom>
            {heading}
          </Typography>
          <FormControl component="fieldset">
            {items.map(item => {
              return (
                <div className={classes.flexRow} key={item.value}>
                  <FormControlLabel
                    label={item.label}
                    control={
                      <Radio
                        checked={this.state.value === item.value}
                        onChange={this.handleChange}
                        value={item.value}
                        name={item.value}
                        aria-label={item.label}
                      />
                    }
                  />
                  <IconButton
                    className={classes.deleteButton}
                    aria-label="Delete"
                    onClick={() => this.deleteItem(item.value)}
                  >
                    <DeleteIcon />
                  </IconButton>
                </div>
              );
            })}
            <Button
              type="submit"
              variant="contained"
              color="primary"
              className={classes.button}
            >
              {locale.SEARCH}
            </Button>
          </FormControl>
        </div>
      </form>
    );
  }
}

FavouriteItems.propTypes = {
  classes: object.isRequired,
};

export default withStyles(styles)(FavouriteItems);
