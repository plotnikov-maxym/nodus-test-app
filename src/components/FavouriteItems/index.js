import {connect} from "react-redux";
import FavouriteItems from "./FavouriteItems";
import {removeFromFavourites} from "../../actions/favourites";
import {fetchArticles, fetchAbstract} from "../../actions/articles";

const mapStateToProps = state => ({
  favourites: state.favourites,
});
const mapDispatchToProps = dispatch => ({
  fetchArticles: query => dispatch(fetchArticles(query)),
  removeFromFavourites: query => dispatch(removeFromFavourites(query)),
  fetchAbstract: id => dispatch(fetchAbstract(id)),
});
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FavouriteItems);
