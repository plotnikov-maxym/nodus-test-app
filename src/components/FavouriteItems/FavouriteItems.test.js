import React from "react";
import renderer from "react-test-renderer";
import {extend} from "underscore";
import {FavouriteItems} from "./FavouriteItems";
import {createShallow} from "@material-ui/core/test-utils";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import FormControl from "@material-ui/core/FormControl";
import Typography from "@material-ui/core/Typography";
import {locale} from "../../constants/locales";

describe("<FavouritesItems />", () => {
  let shallow;

  beforeAll(() => {
    shallow = createShallow();
  });

  describe("FavouritesItems component", () => {
    it("matches snapshot if no FavouritesItems provided", async () => {
      const wrapper = await renderComponent({favourites: {}});
      expect(wrapper).toMatchSnapshot();
    });

    it("matches snapshot with FavouritesItems provided", async () => {
      const wrapper = await renderComponent();
      expect(wrapper).toMatchSnapshot();
    });

    it("renders the correct number of items", async () => {
      const wrapper = await getComponent();
      const props = getMockProps();
      const itemsQty = Object.keys(props.favourites).length;
      expect(wrapper.find(FormControl).children()).toHaveLength(itemsQty);
    });

    it("renders fallback if no favourites provided", async () => {
      const favourites = {};
      const wrapper = await getComponent({favourites});
      const itemsQty = Object.keys(favourites).length;
      expect(wrapper.find(FormControl).children()).toHaveLength(itemsQty);
      const fallbackText = wrapper
        .find(Typography)
        .findWhere(text => text.prop("variant") === "body1");
      expect(fallbackText.props().children).toEqual(locale.NO_FAVOURITES);
    });

    it("renders the correct heading", async () => {
      const heading = "Favourites";
      const wrapper = await getComponent({heading});
      expect(wrapper.find(Typography)).toExist();
      expect(wrapper.find(Typography).props().children).toEqual(heading);
    });

    it("renders the correctly Radio and handles onChange", async () => {
      const favourites = {
        key: "Search",
      };
      const wrapper = await getComponent({favourites});
      const radioBtn = wrapper.find(FormControlLabel);
      expect(radioBtn).toExist();
      expect(radioBtn.prop("control").props.value).toBe(favourites.key);
      expect(wrapper.state("value")).toEqual("");
      radioBtn.prop("control").props.onChange({
        target: {
          value: favourites.key,
        },
      });
      wrapper.update();
      expect(wrapper.state("value")).toEqual(favourites.key);
    });

    it("renders the correctly submit Button and calls proper function on submit", async () => {
      const fetchArticles = jest.fn();
      const favourites = {
        searchQuery: "Search Query",
      };
      const wrapper = await getComponent({favourites, fetchArticles});
      wrapper.setState({value: favourites.searchQuery});
      wrapper.update();
      const submitButton = wrapper
        .find(Button)
        .findWhere(btn => btn.prop("type") === "submit");
      expect(submitButton).toExist();
      const form = wrapper.find("form");
      form.prop("onSubmit")({
        preventDefault: () => {},
      });
      expect(fetchArticles).toHaveBeenCalledWith(favourites.searchQuery);
    });

    it("renders the correctly Delete Button and calls proper function on click and resets state value", async () => {
      const removeFromFavourites = jest.fn();
      const favourites = {
        searchQueryKey: "Search Query",
        searchQueryKey2: "Search Query 2",
      };
      const [deleteKey] = Object.keys(favourites);
      const wrapper = await getComponent({favourites, removeFromFavourites});
      wrapper.setState({value: favourites.searchQueryKey2});
      wrapper.update();
      const deleteBtns = wrapper.find(IconButton);
      expect(deleteBtns).toHaveLength(Object.keys(favourites).length);
      deleteBtns.forEach((btn, index) => {
        btn.simulate("click");
        wrapper.update();
        const favLength = Object.keys(favourites).length;
        // Line 33
        if (index !== favLength - 1) {
          expect(wrapper.state("value")).toEqual(favourites.searchQueryKey2);
        } else {
          expect(wrapper.state("value")).toEqual("");
        }
        expect(removeFromFavourites).toHaveBeenCalledWith(deleteKey);
      });
    });
  });

  const getComponent = props => {
    const parsedProps = extend(getMockProps(), props);

    return shallow(<FavouriteItems {...parsedProps} />);
  };

  const renderComponent = props => {
    const parsedProps = extend(getMockProps(), props);

    return renderer.create(<FavouriteItems {...parsedProps} />).toJSON();
  };

  const getMockProps = () => ({
    classes: {
      button: "FavouriteItems-button-192",
      deleteButton: "FavouriteItems-deleteButton-193",
      flexRow: "FavouriteItems-flexRow-189",
      formControl: "FavouriteItems-formControl-190",
      group: "FavouriteItems-group-191",
      root: "FavouriteItems-root-188",
    },
    removeFromFavourites: () => {},
    fetchArticles: () => {},
    heading: "My Test Heading",
    favourites: {
      MySearchQuery: "My Search Query",
      AnotherTestQueryTest: "Another Test Query - Tessss",
    },
  });
});
