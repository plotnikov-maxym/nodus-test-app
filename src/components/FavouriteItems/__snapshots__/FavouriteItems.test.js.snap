// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavouritesItems /> FavouritesItems component matches snapshot if no FavouritesItems provided 1`] = `
<div
  className="FavouriteItems-root-188"
>
  <div
    className="FavouriteItems-formControl-190"
  >
    <h5
      className="MuiTypography-root-1 MuiTypography-h5-17 MuiTypography-gutterBottom-28"
    >
      My Test Heading
    </h5>
    <p
      className="MuiTypography-root-1 MuiTypography-body1-10 MuiTypography-gutterBottom-28"
    >
      You don't have any favourites at the moment. You can add them at anytime.
    </p>
  </div>
</div>
`;

exports[`<FavouritesItems /> FavouritesItems component matches snapshot with FavouritesItems provided 1`] = `
<form
  autoComplete="off"
  className="FavouriteItems-root-188"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="FavouriteItems-formControl-190"
  >
    <h5
      className="MuiTypography-root-1 MuiTypography-h5-17 MuiTypography-gutterBottom-28"
    >
      My Test Heading
    </h5>
    <div>
      <fieldset
        className="MuiFormControl-root-37"
      >
        <div
          className="FavouriteItems-flexRow-189"
        >
          <label
            className="MuiFormControlLabel-root-41"
          >
            <span
              aria-label="MySearchQuery"
              className="MuiButtonBase-root-62 MuiIconButton-root-56 MuiPrivateSwitchBase-root-52 MuiRadio-root-47 MuiRadio-colorSecondary-51"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label-61"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-65"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <input
                  checked={false}
                  className="MuiPrivateSwitchBase-input-55"
                  disabled={false}
                  name="My Search Query"
                  onChange={[Function]}
                  type="radio"
                  value="My Search Query"
                />
              </span>
              <span
                className="MuiTouchRipple-root-100"
              />
            </span>
            <span
              className="MuiTypography-root-1 MuiTypography-body1-10 MuiFormControlLabel-label-46"
            >
              My Search Query
            </span>
          </label>
          <button
            aria-label="Delete"
            className="MuiButtonBase-root-62 MuiIconButton-root-56 FavouriteItems-deleteButton-193"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-61"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-65"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-100"
            />
          </button>
        </div>
        <div
          className="FavouriteItems-flexRow-189"
        >
          <label
            className="MuiFormControlLabel-root-41"
          >
            <span
              aria-label="AnotherTestQueryTest"
              className="MuiButtonBase-root-62 MuiIconButton-root-56 MuiPrivateSwitchBase-root-52 MuiRadio-root-47 MuiRadio-colorSecondary-51"
              onBlur={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label-61"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-65"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <input
                  checked={false}
                  className="MuiPrivateSwitchBase-input-55"
                  disabled={false}
                  name="Another Test Query - Tessss"
                  onChange={[Function]}
                  type="radio"
                  value="Another Test Query - Tessss"
                />
              </span>
              <span
                className="MuiTouchRipple-root-100"
              />
            </span>
            <span
              className="MuiTypography-root-1 MuiTypography-body1-10 MuiFormControlLabel-label-46"
            >
              Another Test Query - Tessss
            </span>
          </label>
          <button
            aria-label="Delete"
            className="MuiButtonBase-root-62 MuiIconButton-root-56 FavouriteItems-deleteButton-193"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-61"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-65"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-100"
            />
          </button>
        </div>
      </fieldset>
      <button
        className="MuiButtonBase-root-62 MuiButtonBase-disabled-63 MuiButton-root-74 MuiButton-contained-85 MuiButton-containedPrimary-86 MuiButton-raised-88 MuiButton-raisedPrimary-89 MuiButton-disabled-94 FavouriteItems-button-192"
        disabled={true}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="-1"
        type="submit"
      >
        <span
          className="MuiButton-label-75"
        >
          Search
        </span>
      </button>
    </div>
  </div>
</form>
`;
